src/coreComponents/physicsSolvers/SolverBase.cpp

```diff
@@ -604,6 +586,51 @@ bool SolverBase::lineSearchWithParabolicInterpolation( real64 const & time_n,
   return lineSearchSuccess;
 }
 
+real64 SolverBase::assembleResidualOnlyAndCalculateNorm( real64 const & time_n,
+                                                         real64 const & dt,
+                                                         DomainPartition & domain,
+                                                         DofManager const & dofManager,
+                                                         CRSMatrixView< real64, globalIndex const > const & localMatrix,
+                                                         ParallelVector & rhs,
+                                                         string const logMessage )
+{
+  localMatrix.zero();
+  rhs.zero();
+
+  arrayView1d< real64 > const localRhs = rhs.open();
+
+  // The current implementation of this function is incomplete
+  // Ultimately, we want to modify the two calls below to only recompute the residual, and not the Jacobian matrix
+  // At that point, the localMatrix argument of this function will be removed
+
+  assembleSystem( time_n,
```

Comment:
Yes, that would be ideal. I left it for a later PR because it seemed overwhelming, but I agree that it would be a good way to have a unique version of each kernel that could do either 1) Jacobian + residual assembly or just 2) residual assembly. That would save a lot of code and be quite useful in multiple places (line search, convergence check in the outer-loop of sequential schemes, etc)

Reply:
While it's easy to just disable the final step (adding entries into the matrix), the goal IMO should be to also avoid all intermediate evaluations, stack allocations and memory accesses associated with derivatives, which I think is far from trivial. At that point, we should consider if it's easier to just have two copies of each assembly kernel (the residual-only one would be a pretty straightforward stripping down of the full one).